Lab 2 - Adding multiple sort functions

brnbke opened the pull request:
Fixes #398

Review of colleague's PR #308

#### Changes proposed in this PR:
Implementing and testing following sort functions:
Bubble
Insertion 
Mergesort

joel00wood commented on 02_bubble/brnbke/main.go:
```diff
@@ -0,0 +1,91 @@
+	"os"
+)
+
+var out io.Writer = os.Stdout
```
Is the purpose of this purely to simplify testing?

brnbke replied:
Yes pretty much, so I can change the output on subsequent Fprintln to "out" and grab that in the test.

joel00wood commented on 02_bubble/brnbke/main_test.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+var testCases = map[string]struct {
+	input []int
```
Given that these two pieces of the struct are of the same type this could be simplified to a single line

brnbke replied:
Yes, you're right, did not notice that, though not sure what the expected way of doing this is in go TBH.

joel00wood commented on 02_bubble/brnbke/main_test.go:
```diff
@@ -0,0 +1,107 @@
+
+func TestMerge(t *testing.T) {
+	var testCases = map[string]struct {
+		left  []int
```
Similar to the above comment, this could be simplified to `left, right, want []int` as a single line

joel00wood commented on 02_bubble/brnbke/main_test.go:
```diff
@@ -0,0 +1,107 @@
+		t.Run(name, func(t *testing.T) {
+			b := merge(test.left, test.right)
+			result := reflect.DeepEqual(test.want, b)
+			if !result {
```
Is it necessary to assign the reflect response to a value, or is this matter of `gofmt/golangci-lint` eccentricity

brnbke replied:
No good reason that I can see actually, just habit I think

joel00wood reviewed (commented):
Most of my comments are essentially nitpicking with respect to formatting and also some curiosity, looks good to me otherwise.

samuong commented on 02_bubble/brnbke/main.go:
```diff
@@ -0,0 +1,91 @@
+	copy(r, arr)
+
+	for {
+		if !changed {
```
If you moved this if condition to the bottom of the loop body, it would still mean the same thing. But then you could also move the `changed` variable into the scope of the `for` loop.

brnbke replied:
have moved the check as suggested so var now just in the scope of the for loops

samuong commented on 02_bubble/brnbke/main.go:
```diff
@@ -0,0 +1,91 @@
+			if result[j] > tmp {
+				result[j+1], result[j] = result[j], tmp
+			} else {
+				break
```
nitpick: I'd suggest only indenting the `break` block, and having the other one outside the `if` statement. https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

camscale replied:
I agree with Sam. By using:
```
if result[j] <= tmp {
  break
}
result[j+1], result[j] = result[j], tmp
```
It makes it clear what the terminating condition is - I don't need to mentally invert the condition in my head to understand the condition of the special case.

It is also a variation of aligning the happy path to the left edge: https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88 (modulo loop indentation).

samuong commented on 02_bubble/brnbke/main.go:
```diff
@@ -0,0 +1,91 @@
+	return merge(l, r)
+}
+
+func merge(l []int, r []int) []int {
```
You don't need to specify `[]int` for the first parameter.

samuong commented on 02_bubble/brnbke/main.go:
```diff
@@ -0,0 +1,91 @@
+
+func merge(l []int, r []int) []int {
+
+	result := []int{}
```
Given that you know the size of the result (`len(l)+len(r)`), you could create a pre-allocated array here. Then you could guarantee that you never need to regrow the array through a call to `append`.

brnbke replied:
Done

samuong commented on 02_bubble/brnbke/main_test.go:
```diff
@@ -0,0 +1,107 @@
+		test := test
+		var buf bytes.Buffer
+
+		out = &buf
```
What's the purpose of overriding `out` in this test (as well a the others)? I think `main` is the only function that uses it, but let me know if I've missed something.

brnbke replied:
A mistake to be honest, copy paste error, removed.

camscale commented on 02_bubble/brnbke/main.go:
```diff
@@ -0,0 +1,91 @@
+	copy(r, arr)
+
+	for {
+		if !changed {
```
As Sam says, since this condition is the first thing in the loop and it breaks the loop, you may as well put it in the condition of the loop:
```
for changed {
...
}
```
It does read a bit funny as Go doesn't use "while" for this, opting to use "for" for all loops.

brnbke replied:
Moved the check to the bottom of the look to keep the changed var scoped to the loops, not sure I can implement it as "for changed" as need to reset the changed to false each loop and only exit if not changed

camh-anz replied:
You could do:
```
for changed := true; changed; {
    changed = false
    ...
}
```
if you want to keep the var scoped to the loop. But this is fine as is. I won't insist on any further change here.

camscale replied:
[sorry if you get this twice. I wanted to add it in the review, but the comment has gone from the review.]

You could do:
```
for changed := true; changed; {
    changed = false
    ...
}
```
if you want to keep the var scoped to the loop. But this is fine as is. I won't insist on any further change here.

brnbke replied:
Ah I see what you mean, I assume this is the prefered Go way of doing this. will take note for future, but will leave as is for this to avoid further reviews.

camscale replied:
I'm not sure if Go-style really has much to say about this, but I prefer to avoid infinite loops with the terminating condition in the body - it's a bit easier to read and know that the loop will actually terminate. In this particular case it's easy enough to tell, but not always.

camscale commented on 02_bubble/brnbke/main.go:
```diff
@@ -0,0 +1,91 @@
+
+	for i := 1; i < len(result); i++ {
+		tmp := result[i]
+		for j := i - 1; j > -1; j-- {
```
nit: `j > -1` is a little odd. I'd use `j >= 0`, since 0 has meaning but -1 does not (except as being 1 off the meaningful number).

brnbke replied:
Implemented.

@samuong seems you were not being too nitpicky with this suggestion either.

samuong replied:
Hah, good to know I'm not alone! 😄

camscale reviewed (changes requested):
I didn't review the merge sort, just the ones requested for the lab (i ain't got time for that!).

I've chimed in with Sam on a couple of comments which relate to idiomatic Go, so I'm going to request changes for those. The rest you can take or leave (but I suggest taking)

brnbke committed:
Lab 2 Adding 3 sort functions

Adding Bubble, insertion and Merge sort functions

brnbke committed:
Minor changes per code review

removing unneeded buffer out code from tests
cleaning up some non standard syntax

brnbke commented:
Reopening PR as closed in error

camscale reviewed (approved):
LGTM. As per usual, could you either rebase-squash-force-push your changes to remove fixup commits, or just do a squash merge when you merge.

brnbke committed:
Merge branch 'master' into lab2
